### PR TITLE
feat(ui): project remote management

### PR DESCRIPTION
**Pull request by juliendonck**
This integrates the "Manage Remotes" feature.

Closes: #1007 

demo

### CODE REVIEW

**Comment by juliendonck** [comment]
@FintanH The scaffolding is here.

**Comment by juliendonck** [comment]
@FintanH the UI is all here. I'm making some assumptions of data which i'm sure will be wrong. I'm happy to adapt the front end logic once we have an idea what the data will look like. For now this illustrates the goal imo.

File: ui/DesignSystem/Component/Illustration.svelte
Comment (rudolfs) [inline ui/DesignSystem/Component/Illustration.svelte:549]: I wonder how I didn't realise this earlier, but all the illustrations are just emojis from the twemoji set. Which we already have integrated. So we could potentially get rid of all these custom illustration SVGs.

```diff
diff --git a/ui/Modal/ManageRemotes.svelte b/ui/Modal/ManageRemotes.svelte
index a49c5fc..b001d44 100644
--- a/ui/Modal/ManageRemotes.svelte
+++ b/ui/Modal/ManageRemotes.svelte
@@ -1,12 +1,9 @@
 <script lang="ts">
-  import { Variant as IllustrationVariant } from "../src/illustration";
-
   import {
-    Illustration,
     List,
     RemoteListItem,
   } from "../DesignSystem/Component";
-  import { Button, Input } from "../DesignSystem/Primitive";
+  import { Button, Emoji, Input } from "../DesignSystem/Primitive";

   const remotes = [
     {
@@ -56,9 +53,8 @@
 </style>

 <div data-cy="remotes-modal" class="container">
-  <Illustration
-    style="margin-bottom: 1.5rem;"
-    variant={IllustrationVariant.Computer} />
+  <Emoji emoji={'💻'} size="huge" style="margin-bottom: 1.5rem;" />
+
   <h1>Manage remotes</h1>
   <div class="input">
     <Input.Text
```

Emoji
Screenshot 2020-10-12 at 15 27 32

vs

Illustration

Screenshot 2020-10-12 at 15 28 01

Comment (juliendonck) [reply ui/DesignSystem/Component/Illustration.svelte:549]: nice

Comment (juliendonck) [reply ui/DesignSystem/Component/Illustration.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/1014/commits/903c0e71db7234671546ea6ece63c1cebdb3dee9

Comment (juliendonck) [reply ui/DesignSystem/Component/Illustration.svelte]: Ended up using the <Illustration> component which is then in turn using the <Emoji> component. Made more sense like this.

**Review by FintanH — COMMENTED** [review]
Looks cool! One thought below on data representation.

I'm also wondering if the "Follow" disabling is going to happen in the next iteration of this?

File: ui/Modal/ManageRemotes.svelte
Comment (xla) [reply ui/Modal/ManageRemotes.svelte:17]: What's the meaning of being synced with a peer? Couldn't imagine how we determine that at the moment.

Comment (FintanH) [reply ui/Modal/ManageRemotes.svelte:17]: The idea behind it is that we have tracked them but haven't got their changes in our repo yet. So it's determined by having them tracked for that project, but no remote exists for them under that project.

Comment (xla) [reply ui/Modal/ManageRemotes.svelte:17]: Might be a matter of wording then. We syncing as a mechanism in another way already.

Comment (xla) [reply ui/Modal/ManageRemotes.svelte:17]: > but maybe is it too early to introduce the TS stuff? I'm not sure what the dev flow of UI components is :)

The component is already TS so no harm in modelling the data.

Comment (FintanH) [reply ui/Modal/ManageRemotes.svelte:17]: > Might be a matter of wording then. We syncing as a mechanism in another way already.

Ya maybe something better is `type RemoteStatus = Tracking | Searching`. And actually, to refine the type above, it would be better represented by:
```ts
interface Tracked {
  kind: "tracked";
  handle: string;
  peerId: string;
  maintainer: boolean;
}

interface Searching {
  kind: "searching";
  peerId: string;
}

type RemoteStatus = Tracked | Searching;
```

Since we know nothing of the `Searching` variant other than the `peerId`, which gets rid of these **nasty** `null`s

**Comment by juliendonck** [comment]
> I'm also wondering if the "Follow" disabling is going to happen in the next iteration of this?

Good question, could we already do it in this iteration? Like check if the peerID is the maintainer's and user's peerID? 
It's hard for me to say since I don't know what data we have available.

**Comment by FintanH** [comment]
> Good question, could we already do it in this iteration? Like check if the peerID is the maintainer's and user's peerID?

So to me, it looks like the data model you came up with makes sense (with a few tweaks I mentioned above ;)). So couldn't we use this maintainer field to disallow the button?

I guess one thing I should have questioned, why would we even show ourselves in the list here as a user? Like, it's obvious that we have the project because we wouldn't be able to click this button otherwise. That would simplify the above to: "we can't unfollow maintainers".

File: ui/DesignSystem/Component/UntrackedRemoteListItem.svelte
Comment (xla) [inline ui/DesignSystem/Component/UntrackedRemoteListItem.svelte:18]: Same as above.

File: ui/DesignSystem/Component/TrackedRemoteListItem.svelte
Comment (xla) [inline ui/DesignSystem/Component/TrackedRemoteListItem.svelte:16]: Should be a dispatch this component should not do any data manipulation, ought to be responsibility of the modal.

File: ui/DesignSystem/Component/TrackedRemoteListItem.svelte
Comment (xla) [inline ui/DesignSystem/Component/TrackedRemoteListItem.svelte:9]:
```suggestion
  import { BadgeType } from "../../src/badge.ts";
  import { PeerType, removeRemote, Role } from "../../src/project";
  import { CSSPosition } from "../../src/style";
  
  import { Avatar, Flex } from "../../DesignSystem/Primitive";
  
  import Badge from "./Badge.svelte";
  import Tooltip from "./Tooltip.svelte";
  import TrackToggle from "./TrackToggle.svelte";

```

File: ui/DesignSystem/Component/TrackedRemoteListItem.svelte
Comment (xla) [inline ui/DesignSystem/Component/TrackedRemoteListItem.svelte:1]:
```suggestion
<script lang="typescript">
```

File: ui/DesignSystem/Component/TrackedRemoteListItem.svelte
Comment (xla) [inline ui/DesignSystem/Component/TrackedRemoteListItem.svelte:13]:
```suggestion
  export let peer: Peer = null;
  export let projectUrn: Urn = null;
  export let projectName: string = null;
```

File: ui/DesignSystem/Component/UntrackedRemoteListItem.svelte
Comment (xla) [inline ui/DesignSystem/Component/UntrackedRemoteListItem.svelte:1]:
```suggestion
<script lang="typescript">
```

File: ui/DesignSystem/Component/UntrackedRemoteListItem.svelte
Comment (xla) [inline ui/DesignSystem/Component/UntrackedRemoteListItem.svelte:9]:
```suggestion
  export let peer: Peer = null;
  export let projectName: string = null;
  export let projectUrn: Urn = null;
```

File: ui/DesignSystem/Component/UntrackedRemoteListItem.svelte
Comment (xla) [inline ui/DesignSystem/Component/UntrackedRemoteListItem.svelte:5]:
```suggestion
  import { removeRemote } from "../../src/project";
  
  import { Flex } from "../../DesignSystem/Primitive";
  import TrackToggle from "./TrackToggle.svelte";
```

**Review by xla — CHANGES_REQUESTED** [review]
The main thing to change is that we start to use remotes as concept even so it's the management of peers in the context of the peer selector. It would streamline if we use peers consistently.

**Comment by rudolfs** [comment]
> The main thing to change is that we start to use remotes as concept even so it's the management of peers in the context of the peer selector. It would streamline if we use peers consistently.

Just to make sure I understand this correctly, you're suggesting we use to `peer` in code and show `remote` in the UI?

**Comment by xla** [comment]
> @xla: just to make sure I understand this correctly, you're suggesting we use `peer` in code and show `remote` in the UI?

I'm surprised that we show remote in the UI as well, while it is the technically correct in the sense of the git file tree layout, it's peers we are referring to, supported by the fact that the remotes are keyed by peer id and it's waht we provide when following new views of a project.

File: ui/DesignSystem/Component/UntrackedRemoteListItem.svelte
Comment (rudolfs) [reply ui/DesignSystem/Component/UntrackedRemoteListItem.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/1014/commits/1385636c46b3c1812fd6233db5b8731b57143038

File: ui/DesignSystem/Component/TrackedRemoteListItem.svelte
Comment (rudolfs) [reply ui/DesignSystem/Component/TrackedRemoteListItem.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/1014/commits/1385636c46b3c1812fd6233db5b8731b57143038

**Comment by cloudhead** [comment]
We're adding remotes to the repository (which happen to be keyed by PeerId), hence why we're calling it remotes.

**Comment by xla** [comment]
> We're adding remotes to the repository (which happen to be keyed by PeerId), hence why we're calling it remotes. Adding "peers" doesn't really having a meaning to the user. What's a peer? What does it do? What does adding a peer mean?

I doubt adding remotes has more significance.

**Review by cloudhead — APPROVED** [review]
Awesome.

The only thing I'd change is the copy on the button: I think it could simply say `Add` or `Follow`.

**Review by xla — CHANGES_REQUESTED** [review]
Can't wait for this to land. Had a first pass with some comment sinline.

File: ui/src/project.ts
Comment (xla) [inline ui/src/project.ts:496]: Let's make sure this regex ony matches valid `base32`.

File: ui/DesignSystem/Component/FollowButton.svelte
Comment (rudolfs) [reply ui/DesignSystem/Component/FollowButton.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/1014/commits/db394c30a83968aa320cf2dbf5887a62290bcada

File: ui/src/track.ts
Comment (rudolfs) [reply ui/src/track.ts:1]: Yeah, we were only using this type in `FollowToggle` and we couldn't use the types at the call-site, because of special svelte syntax for the handlers like `on:eventname`. `on:Event.Track` doesn't compile...

So I deleted this file:

Screenshot 2020-10-27 at 15 24 31

File: ui/src/project.ts
Comment (FintanH) [reply ui/src/project.ts:496]: Nooooo base58 hahaha https://github.com/radicle-dev/radicle-link/blob/master/spec/radicle-link.md#cryptographic-identities

File: ui/Modal/ManageRemotes.svelte
Comment (rudolfs) [reply ui/Modal/ManageRemotes.svelte]: Good catch!

https://github.com/radicle-dev/radicle-upstream/pull/1014/commits/a20dd25a5733d5183189118e06af39804154ffd6

File: ui/src/project.ts
Comment (xla) [reply ui/src/project.ts:496]: What @FintanH said.

File: ui/Screen/Profile/Following.svelte
Comment (rudolfs) [reply ui/Screen/Profile/Following.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/1014/commits/a20dd25a5733d5183189118e06af39804154ffd6

File: ui/Modal/ManageRemotes.svelte
Comment (FintanH) [inline ui/Modal/ManageRemotes.svelte:28]: Just to check, does the `peerValidation.reset()` above cover this case?

File: ui/Modal/ManageRemotes.svelte
Comment (FintanH) [inline ui/Modal/ManageRemotes.svelte:80]:
```suggestion
    <h1>Manage Peers</h1>
```

I think capital P makes sense since it's a header but also happy with "peers"

File: ui/Modal/ManageRemotes.svelte
Comment (FintanH) [inline ui/Modal/ManageRemotes.svelte:1]: Should this file be called `ManagePeers.svelte`?

File: ui/Modal/ManageRemotes/Peer.svelte
Comment (FintanH) [inline ui/Modal/ManageRemotes/Peer.svelte:1]: And this directory `ManagePeers/*`?

File: ui/Modal/ManageRemotes/PeerFollowRequest.svelte
Comment (FintanH) [inline ui/Modal/ManageRemotes/PeerFollowRequest.svelte:16]: This logic exists elsewhere as well, right? Should we put it in a shared location?

File: ui/Modal/ManageRemotes.svelte
Comment (rudolfs) [reply ui/Modal/ManageRemotes.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/1014/commits/53efd607ff814c41c341fae044ad1f2d282ff55d

File: ui/Modal/Search.svelte
Comment (FintanH) [inline ui/Modal/Search.svelte:20]: Should the component be called `Peer`?

File: ui/src/project.ts
Comment (FintanH) [inline ui/src/project.ts:201]: Should this be `FollowPeer`? :eyes:

File: ui/src/project.ts
Comment (FintanH) [inline ui/src/project.ts:498]: Just so you know, track and untrack are idempotent so they won't double track or anything. But if this is for UX then totaly ignore.

**Review by FintanH — COMMENTED** [review]
We should definitely make an effort to make more sub-modules -- `project.ts` is blowing up with different concepts.

Nice work :clap: Feel free to dismiss some of my feedback if it's not relevant.

File: ui/DesignSystem/Component/List.svelte
Comment (rudolfs) [reply ui/DesignSystem/Component/List.svelte:9]: I see where you're coming from. I also didn't like this approach.

The thing is, it doesn't make sense to extract the `<li>` styling outside of this component, because we want the list elements to look the same. So it makes sense that the styles are encapsulated in `<List>`.

What we could do is have two specialised components, instead of one generic: `<List>` and `<ClickableList>` for example.

File: ui/src/project.ts
Comment (rudolfs) [reply ui/src/project.ts]: https://github.com/radicle-dev/radicle-upstream/pull/1014/commits/b5098322f2d4453b6229c85827c57814645aae52

File: ui/Modal/ManageRemotes.svelte
Comment (rudolfs) [reply ui/Modal/ManageRemotes.svelte:28]: Correct, it does.

File: ui/Modal/Search.svelte
Comment (rudolfs) [reply ui/Modal/Search.svelte:20]: This `Remote` component is actually not related to our domain. It already existed in our system and is for remote data fetching. 🤷

File: ui/src/project.ts
Comment (rudolfs) [reply ui/src/project.ts:201]: Only if we go all the way and also rename the proxy endpoints. Not sure it's worth the effort at the moment.

File: ui/src/project.ts
Comment (rudolfs) [reply ui/src/project.ts:498]: Thanks, yeah, was aware of that. It's only for UX.

File: ui/Modal/ManageRemotes.svelte
Comment (rudolfs) [reply ui/Modal/ManageRemotes.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/1014/commits/8e5e847a1b0d3569a02b867b97c9b60a578f7f19

File: ui/Modal/ManageRemotes/Peer.svelte
Comment (rudolfs) [reply ui/Modal/ManageRemotes/Peer.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/1014/commits/8e5e847a1b0d3569a02b867b97c9b60a578f7f19

File: ui/Modal/ManageRemotes.svelte
Comment (rudolfs) [reply ui/Modal/ManageRemotes.svelte]: See: https://github.com/radicle-dev/radicle-upstream/pull/1014#issuecomment-717222964.

File: ui/Modal/ManageRemotes/PeerFollowRequest.svelte
Comment (rudolfs) [reply ui/Modal/ManageRemotes/PeerFollowRequest.svelte]: I simplifier it to a one-liner. I think the duplication is okay for now.

**Comment by rudolfs** [comment]
> We should definitely make an effort to make more sub-modules -- project.ts is blowing up with different concepts.

@FintanH yeah, I agree `project.ts` is getting out of hand, however I think we should attempt the refactoring in a separate PR. I bumped into circular dependency issues previously when trying to split this up.

**Review by FintanH — APPROVED** [review]
Nice work @rudolfs and @juliendonck :grin:

**Review by FintanH — APPROVED** [review]
../approve

**Review by FintanH — APPROVED** [review]
`assert!(has_tests, true)`

**Review by cloudhead — APPROVED** [review]

**Pull request merged**
